File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer token;

    static List<int[]> paths;
    static int[] parent;
    static long res = 0;

    public static void main(String[] args) throws IOException {
        token = new StringTokenizer(input.readLine());
        int v = Integer.parseInt(token.nextToken());
        int e = Integer.parseInt(token.nextToken());

        paths = new ArrayList<>();
        for (int i = 0; i < e; i++) {
            token = new StringTokenizer(input.readLine());
            paths.add(new int[]{Integer.parseInt(token.nextToken()), Integer.parseInt(token.nextToken()), Integer.parseInt(token.nextToken())});
        }
        paths.sort((o1, o2) -> {
            if (o1[2] == o2[2]) {
                if (o1[1] == o2[1]) {
                    return o1[0] - o2[0];
                }
                return o1[1] - o2[1];
            }
            return o1[2] - o2[2];
        });

        parent = new int[v+1];
        for (int i = 0; i< parent.length; i++) {
            parent[i] = i;
        }

        for (int[] path : paths) {
            int a = path[0];
            int b = path[1];
            int c = path[2];

            if (find(a)==find(b)) continue;
            union(a,b);
            res+=c;
        }

        System.out.println(res);
    }
    public static int find(int x){
        if (x == parent[x]) return x;
        return parent[x] = find(parent[x]);
    }
    public static void union(int x, int y){
        x = find(x);
        y = find(y);
        if (x == y) return;
        parent[y] = x;
    }
}

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
'''
아이디어:
MST 유니온 파인드로 구현하기
'''
import sys

sys.setrecursionlimit(10 ** 5)


def find(x):
    if parent[x] != x:
        parent[x] = find(parent[x])  # 경로 압축
    return parent[x]


def union(a, b):
    a = find(a)
    b = find(b)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b


N, M = map(int, sys.stdin.readline().split())
parent = []
for i in range(N + 1):
    parent.append(i)

edges = []
for _ in range(M):
    a, b, c = map(int, sys.stdin.readline().split())
    edges.append((c, a, b))
edges.sort() # 가중치 낮은 순부터 시도해야해서

res = 0
for edge in edges:
    w, a, b = edge
    if find(a) != find(b): # 같은 그룹이 아니면
        union(a, b)
        res += w

print(res)
